Mathlib/Analysis/Normed/Unbundled/SpectralNorm.lean

```diff
@@ -644,4 +664,218 @@ theorem spectralAlgNorm_isPowMul : IsPowMul (spectralAlgNorm K L) := isPowMul_sp
+      add_le' := fun a b => map_add_le_add _ _ _
+      neg' := fun y => by simp [(algebraMap K⟮x⟯ L).map_neg y]
+      mul_le' := fun a b => map_mul_le_mul _ _ _
+      eq_zero_of_map_eq_zero' := fun a ha => by
+        simpa [map_eq_zero_iff_eq_zero, map_eq_zero] using ha }
+  letI n2 : NormedRing K⟮x⟯ := RingNorm.toNormedRing hf_norm
+  letI N2 : NormedSpace K K⟮x⟯ :=
+    { one_smul := fun e => by simp [one_smul]
+      mul_smul  := fun k1 k2 e => by simp [mul_smul]
+      smul_zero := fun e => by simp
+      smul_add  := fun k e_1 e_2 => by simp [smul_add]
+      add_smul  := fun k_1 k_2 e => by simp [add_smul]
+      zero_smul := fun e => by simp [zero_smul]
+      norm_smul_le := fun k y => by
+        change (f ((algebraMap K⟮x⟯ L) (k • y)) : ℝ) ≤ ‖k‖ * f (algebraMap K⟮x⟯ L y)
+        have : (algebraMap (↥K⟮x⟯) L) (k • y) = k • algebraMap (↥K⟮x⟯) L y := by
+          simp [IntermediateField.algebraMap_apply]
+        rw [this, map_smul_eq_mul] }
+  haveI hKx_fin : FiniteDimensional K ↥K⟮x⟯ :=
+    IntermediateField.adjoin.finiteDimensional (Algebra.IsAlgebraic.isAlgebraic x).isIntegral
+  haveI : FiniteDimensional K E := hKx_fin
+  set Id1 : K⟮x⟯ →L[K] E := ⟨id1, id1.continuous_of_finiteDimensional⟩
+  set Id2 : E →L[K] K⟮x⟯ := ⟨id2, id2.continuous_of_finiteDimensional⟩
+  obtain ⟨C1, hC1_pos, hC1⟩ : ∃ C1 : ℝ, 0 < C1 ∧ ∀ y : K⟮x⟯, ‖id1 y‖ ≤ C1 * ‖y‖ :=
+    Id1.isBoundedLinearMap.bound
+  obtain ⟨C2, hC2_pos, hC2⟩ : ∃ C2 : ℝ, 0 < C2 ∧ ∀ y : E, ‖id2 y‖ ≤ C2 * ‖y‖ :=
+    Id2.isBoundedLinearMap.bound
+  exact ⟨ C2, C1, hC2_pos, hC1_pos,
+    forall_and.mpr ⟨fun y ↦ hC2 ⟨y, (IntermediateField.algebra_adjoin_le_adjoin K _) y.2⟩,
+      fun y ↦ hC1 ⟨y, (IntermediateField.algebra_adjoin_le_adjoin K _) y.2⟩⟩⟩
+
+/-- If `K` is a field complete with respect to a nontrivial nonarchimedean multiplicative norm and
+  `L/K` is an algebraic extension, then any multiplicative ring norm on `L` extending the norm on
+  `K` coincides with the spectral norm. -/
+theorem spectralNorm_unique_field_norm_ext [CompleteSpace K]
+    {f : AbsoluteValue L ℝ} (hf_ext : ∀ (x : K), f (algebraMap K L x) = ‖x‖) (x : L) :
+    f x = spectralNorm K L x := by
+  set g : AlgebraNorm K L :=
+    { MulRingNorm.mulRingNormEquivAbsoluteValue.invFun f with
+      smul' := fun k x => by
+        simp only [AddGroupSeminorm.toFun_eq_coe, MulRingSeminorm.toFun_eq_coe]
+        rw [Algebra.smul_def, map_mul]
+        congr
+        rw [← hf_ext k]
+        rfl
+      mul_le' := fun x y => by
+        simp [AddGroupSeminorm.toFun_eq_coe, MulRingSeminorm.toFun_eq_coe] }
+  have hg_pow : IsPowMul g := MulRingNorm.isPowMul _
+  have hgx : f x = g x := rfl
+  rw [hgx, spectralNorm_unique hg_pow, spectralAlgNorm_def]
+
+/-- Given a nonzero `x : L`, the corresponding `seminormFromConst` can be regarded as an algebra
+  norm, when one assumes that `(spectralAlgNorm h_alg hna) 1 ≤ 1`. -/
+def algNormFromConst (h1 : (spectralAlgNorm K L).toRingSeminorm 1 ≤ 1) {x : L} (hx : x ≠ 0) :
```

Comment:
When trying to understand this work, this docstring was kind of useless because I didn't know what `seminormFromConst` was, and then the docstring for `seminormFromConst` was also no help because it is basically tautological; I finally found `seminormFromConst'` whose docstring told me what this function was! Maybe this docstring should explicitly say the formula for the seminorm?